### PR TITLE
AMQP connection recovery creates extra connection/channel.

### PR DESCRIPTION
**Pull request by vishnu-alapati**
### Description of the Change

Made the necessary changes to resolve extra AMQP connection issue.

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by: Vishnu Alapati vishnu.alapati@tcs.com

### CODE REVIEW

File: publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/helper/RabbitMqProperties.java [hunk 1]
```diff
@@ -272,15 +273,36 @@ public void createRabbitMqConnection() throws RemRemPublishException {
             }
             factory.setConnectionTimeout(tcpTimeOut);
             rabbitConnection = factory.newConnection();
+            rabbitConnection.addShutdownListener(new ShutdownListener() {
+                @Override
+                public void shutdownCompleted(ShutdownSignalException cause) {
+                    log.debug("Connection Shutdown completed " + cause);
+                    try {
```
Comment (z-sztrom) [inline publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/helper/RabbitMqProperties.java:280]: I think the section is useless. The connection is always closed at that point of time...

File: publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/helper/RabbitMqProperties.java [hunk 2]
```diff
@@ -272,15 +273,36 @@ public void createRabbitMqConnection() throws RemRemPublishException {
             }
             factory.setConnectionTimeout(tcpTimeOut);
             rabbitConnection = factory.newConnection();
+            rabbitConnection.addShutdownListener(new ShutdownListener() {
+                @Override
+                public void shutdownCompleted(ShutdownSignalException cause) {
+                    log.debug("Connection Shutdown completed " + cause);
```
Comment (z-sztrom) [inline publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/helper/RabbitMqProperties.java:279]: We should probably use cause.getMessage(). We don't need the whole stacktrace.

File: publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/helper/RabbitMqProperties.java [hunk 3]
```diff
@@ -272,15 +273,36 @@ public void createRabbitMqConnection() throws RemRemPublishException {
...
+                        // TODO Auto-generated catch block
+                        e.printStackTrace();
+                    }
+                }
+            });
+
+            rabbitConnection.addBlockedListener(new BlockedListener() {
+                public void handleBlocked(String reason) throws IOException {
+                    // Connection is now blocked
+                    log.debug("Connection is blocked " + reason);
```
Comment (z-sztrom) [inline publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/helper/RabbitMqProperties.java:292]: reason.getMessage()

File: publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/helper/RabbitMqProperties.java [hunk 4]
```diff
@@ -289,6 +311,32 @@ public void createRabbitMqConnection() throws RemRemPublishException {
...
+     */
+    private Channel createNewChannel() throws IOException {
+        Channel channel = rabbitConnection.createChannel();
+        channel.addShutdownListener(new ShutdownListener() {
+            public void shutdownCompleted(ShutdownSignalException cause) {
+                // Beware that proper synchronization is needed here
+                if (cause.isInitiatedByApplication()) {
+                    log.debug("Shutdown is initiated by application. Ignoring it.");
+                } else {
+                    log.error("Shutdown is NOT initiated by application." + rabbitConnection);
```
Comment (z-sztrom) [inline publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/helper/RabbitMqProperties.java:326]: Add a space after the dot '.' to separate message from connection description.

File: publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/helper/RabbitMqProperties.java (hunk 3 above)
Comment (vishnu-alapati) [reply publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/helper/RabbitMqProperties.java:292]: the variable reason is type of string, which returns the proper message.

File: publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/helper/RabbitMqProperties.java (hunk 2 above)
Comment (vishnu-alapati) [reply publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/helper/RabbitMqProperties.java:279]: Ok

File: publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/helper/RabbitMqProperties.java (hunk 1 above)
Comment (vishnu-alapati) [reply publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/helper/RabbitMqProperties.java:280]: I have tested by removing this piece of code than I'm seeing more connections..,

Comment (z-sztrom) [reply publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/helper/RabbitMqProperties.java:280]: Hmm, very interesting... In that case we should change throwable to IOException or AlreadyClosedException or something what is thrown. Next, e.printStackTrace() should be removed---in fact we don't need any code there. Instead add a comment explainig why we have this try/catch block and that should not be removed.

Comment (vishnu-alapati) [reply publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/helper/RabbitMqProperties.java:280]: Sure Roman

**Review by z-sztrom — APPROVED** [review]

**Review by piyush-sadangi — APPROVED** [review]
LGTM

**Review by z-sztrom — APPROVED** [review]
No more extra connections observed during testing.

**Pull request merged**